java-squid/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java

```diff
@@ -358,12 +358,21 @@ public void visitParameterizedType(ParameterizedTypeTree tree) {
 
   @Override
   public void visitWildcard(WildcardTree tree) {
-    if (tree.bound() == null) {
-      registerType(tree, Symbols.unknownType);
-    } else {
+    JavaType bound = symbols.objectType;
+    boolean unbounded = tree.is(Tree.Kind.UNBOUNDED_WILDCARD);
+    if (!unbounded) {
       resolveAs(tree.bound(), JavaSymbol.TYP);
-      registerType(tree, getType(tree.bound()));
+      bound = getType(tree.bound());
+    }
+    registerType(tree, newWildcardType(tree, bound, unbounded));
+  }
+
+  private JavaType.WildCardType newWildcardType(WildcardTree tree, JavaType bound, boolean unbunded) {
+    String name = unbunded ? "?" : "? " + tree.extendsOrSuperToken().text() + " " + tree.bound().symbolType().name();
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add parentheses to make the operator precedence explicit. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS864)